Add documentation for upgrading to 2.0.0 / 6.0.0 of Chocolatey products (#722)
chocolatey/docs: input/en-us/guides/migrations/upgrading-to-chocolatey-v2-v6.md

The release notes for each product is listed below:

* [Chocolatey CLI](https://docs.chocolatey.org/en-us/choco/release-notes#may-31-2023)
* [Chocolatey Licensed Extension](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#may-31-2023)
* [Chocolatey Agent](https://docs.chocolatey.org/en-us/agent/release-notes#may-31-2023)
* [Chocolatey GUI](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#may-31-2023)
* [Chocolatey GUI Licensed Extension](https://docs.chocolatey.org/en-us/chocolatey-gui-licensed-extension/release-notes#may-31-2023)

## .NET Framework 4.8 Required

Alongside the NuGet v3 Feed Support, Chocolatey CLI now requires .NET Framework 4.8.

When Chocolatey CLI is installed, the installation process will try to install .NET Framework 4.8 if it's not already present on the system. This will require a reboot before Chocolatey CLI can be installed and means that the installation process will have to be started again.

We recommend you install .NET Framework 4.8 and reboot before installing or upgrading to Chocolatey CLI v2.0.0.

> If you are unable to upgrade to .NET Framework 4.8, **we do not recommend** you upgrade to Chocolatey CLI version 2.0.0.

Reviewer: Should we include a warning that only licensed users can get support for `1.x`?

That any new issues created for the previous major version by OSS users will not be accepted?

Author: Added a link to the Support Lifecycle docs.

### NuGet v3 Feed Support Added

Chocolatey CLI has supported NuGet v2 feeds since it's earliest days. But, like all things, time and technology marches on and NuGet v3 feeds have a new way of working and new opportunities for package management. Chocolatey CLI now supports NuGet v3 package feeds and listed below are some of the repository managers that also support them:

* [GitHub Packages](https://github.com/features/packages)
* [ProGet](https://inedo.com/proget)
* [Sonatype Nexus](https://www.sonatype.com/products/sonatype-nexus-repository)
* [JFrog Artifactory](https://jfrog.com/artifactory/)

Adding Nuget v3 feed support to Chocolatey CLI has been a long-term goals that we are delighted to have completed. It required a substantial development and is a major change to the Chocolatey CLI internals.

## SemVer 2.0.0 Support

> We have addded support for [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) that allows previously invalid pre-release version numbering such as `1.0.1-alpha.23`. The only constraint is that the version number segments fit into the range `0` to `2147483647`.

Reviewer: Should we mention that CCR won't be supporting SemVer v2 for the time being?
And SemVer v2 should only be used for local or custom/internal sources?

Running `choco list --local-only` or `choco list -lo` will now show this message:

```
Invalid argument --local-only. This argument has been removed from the list command and cannot be used.
```

To ensure we do not break any existing automation or scripts using the `--limit-output` or `-r` options, we add a warning to the logs but otherwise ignore the options. As this will be removed in a future release, please ensure remove the `--local-only` and `-lo` options fron any automations or scripts that you have.

### Chocolatey Shortcut Shims Removed

In addition to the [Chocolatey shortcut shims removed in version 1.0.0](https://github.com/chocolatey/choco/issues/2468), we have removed the [remaining ones](https://github.com/chocolatey/choco/issues/2641).

The shims removed (inlcuding the earlier ones) and their equivalent full commands are:

* `chocolatey`: use `choco`.
* `cinst`: use `choco install`.
* `clist`: use `choco list`.
* `cpush`: use `choco push`.
* `cuninst`: use `choco uninstall`.
* `cup`: `choco upgrade`.

> * `cver`: `choco version`.

Author: I was getting confused as I knew it was in there. It's now `choco --version`. Updated it.

> If you want to add these shortcuts back, you can can create a PowerShell function or batch file to provide the same functionality.

Author: No :) I deliberately strayed away from that. I'll add some though.

## Modified API Key Commands

To make working with Chocolatey CLI consistent, we have changed the `choco apikey` command to add the following subcommands:

* `add`: add an API key.
* `remove`: remove an API key.
* `list`: list the API keys.

As the `remove` subcommand has been added, we have removed the `--remove` options.

## What To Do If Uprading Fails

If you follow the steps below and run into issues that you are unable to resolve, support is available:

* **Licensed Customers**. Please run `choco support` to find the support options available to you.
* **Free and Open-Source (FOSS) Users**. Community support is available on [Discord](https://ch0.co/community).

## Upgrading Chocolatey Products

> :choco-warning: **WARNING**
>
> You **must** upgrade Chocolatey CLI to 1.4.0 before you start your upgrade to the Chocolatey CLI 2.0.0 or Chocolatey Licensed Extension 6.0.0 versions. Failure to do this may result in Chocolatey products being non-functional.

The Chocolatey products are inter-dependent on each other and all of the products must be upgraded. You cannot upgrade one product and leave the others at older vesions.

### Latest 1.x and 5.x Stable Chocolatey Products

> To upgrade, you need to ensure you are at the latest stable version of the Chocolatey products which are:

Reviewer: Do you think we should mention that this is only required for licensed customers with more than just CLI, GUI, and CLE installed?

If just those three products are installed, it should work without any problems without needing to upgrade all products.

Author: I was trying to stay away from having multiple options when upgrading. They need to be on the latest versions for all the product they use. I might be misreading what you're asking, but are you asking me to make it clearer they only need to be on the latest version for the products _they use_?

> :choco-warning: **WARNING**
>
> If you need to specify a `--source` option in the Advanced Deployment PowerShell script, please make sure you reference only one source value. For example, `--source="'my-internal-repo'"` and not `--source="'my-internal-repo;chocolatey.licensed'"`. Specifying multiple sources for this specific upgrade may cause the Chocolatey product packages to not be found. Note that this only affects the upgrade to version 2.0.0 or 6.0.0 of the Chocolatey products and not future packages you may install or upgrade.

Once you have completed step 1 above, you will be ready to start the upgrade to the latest version of Chocolatey products:

1. Using the same group that used in Step 1 above, create an [Advanced Deployment](xref:ccm-deployments) with two Steps:

  1. The first Deployment Step will upgrade Chocolatey Agent through a Windows scheduled task. Add the [recommended code](xref:upgrade-agent#using-chocolatey-central-management-to-upgrade-chocolatey-agent) as the PowerShell script. If you use an internal repository, change the `--source 'chocolatey.licensed'` in the command, to your internal repository location.

>   1. The second Deployment Step ensures that all other Chocolatey products are now at their latest stable 2.0.0 and 6.0.0 versions. The Advanced Deployment Powershell script should be `choco upgrade chocolatey --source="'chocolatey.licensed'"`. If you use an internal repository, change the `--source="'chocolatey.licensed'"` in the command, to your internal repository location.

Author: Yes. I'll add it up at the top (under Upgrading Considerations). Please review that to ensure I'm correct.

>   1. The second Deployment Step ensures that all other Chocolatey products are now at their latest stable 2.0.0 and 6.0.0 versions. The Advanced Deployment Powershell script should be `choco upgrade chocolatey --source="'chocolatey.licensed'"`. If you use an internal repository, change the `--source="'chocolatey.licensed'"` in the command, to your internal repository location.

Author: I've added that above.

We recommend you install .NET Framework 4.8 and reboot before installing or upgrading to Chocolatey CLI v2.0.0.

If you are unable to upgrade to .NET Framework 4.8, **we do not recommend** you upgrade to Chocolatey CLI version 2.0.0.

### NuGet v3 Feed Support Added

Chocolatey CLI has supported NuGet v2 feeds since it's earliest days. But, like all things, time and technology marches on and NuGet v3 feeds have a new way of working and new opportunities for package management. Chocolatey CLI now supports NuGet v3 package feeds and listed below are some of the repository managers that also support them:

* [GitHub Packages](https://github.com/features/packages)
* [ProGet](https://inedo.com/proget)
* [Sonatype Nexus](https://www.sonatype.com/products/sonatype-nexus-repository)
* [JFrog Artifactory](https://jfrog.com/artifactory/)

Adding Nuget v3 feed support to Chocolatey CLI has been a long-term goals that we are delighted to have completed. It required a substantial development and is a major change to the Chocolatey CLI internals.

## SemVer 2.0.0 Support

> We have addded support for [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) that allows previously invalid pre-release version numbering such as `1.0.1-alpha.23`. The only constraint is that the version number segments fit into the range `0` to `2147483647`.

Reviewer 2: ```suggestion
We have added support for [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) that allows previously invalid pre-release version numbering such as `1.0.1-alpha.23`. The only constraint is that the version number segments fit into the range `0` to `2147483647`.

```

## Minimum Operating System Support Changes

Please see our [Support Lifecycle](xref:support-lifecycle) documentation for more information.

### Side-By-Side Installs Have Been Removed

> Side-by-side installs were introduced in Chocolatey CLI to help with the challenge of installing multiple versions of the same package. As use of the functionality grew and expanded beyond the bondaries of the narrow use case that it was intended to help with, so to did the number of users running into issues with it. The use of side-by-side installs began to cause more problems that it solved and the decison was made to remove it from version 2.0.0 of Chocolatey CLI.

Reviewer 2: ```suggestion
Side-by-side installs were introduced in Chocolatey CLI to help with the challenge of installing multiple versions of the same package. As use of the functionality grew and expanded beyond the boundaries of the narrow use case that it was intended to help with, so to did the number of users running into issues with it. The use of side-by-side installs began to cause more problems that it solved and the decision was made to remove it from version 2.0.0 of Chocolatey CLI.

```

We do have plans to implement a similar feature in future that will allow multiple versions of a package to be installed, but do not have any ny timescale for that as yet.

If you need side-by-side functionality, **we do not recommend** you upgrade to Chocolatey CLI version 2.0.0.

### The List Command Now Lists Local Packages Only and the `--local-only` and `-lo` Options Have Been Removed

In version 1.0.0 of Chocolatey CLI, we added notices that the `choco list` command will list only local packages, and deprecated the `--local-only` and `-lo` options. See this [GitHub issue for more information](https://github.com/chocolatey/choco/issues/158).

Running `choco list --local-only` or `choco list -lo` will now show this message:

```
Invalid argument --local-only. This argument has been removed from the list command and cannot be used.
```

> To ensure we do not break any existing automation or scripts using the `--limit-output` or `-r` options, we add a warning to the logs but otherwise ignore the options. As this will be removed in a future release, please ensure remove the `--local-only` and `-lo` options fron any automations or scripts that you have.

Reviewer 2: ```suggestion
To ensure we do not break any existing automation or scripts using the `--limit-output` or `-r` options, we add a warning to the logs but otherwise ignore the options. As this will be removed in a future release, please ensure remove the `--local-only` and `-lo` options from any automations or scripts that you have.

```

### Chocolatey Shortcut Shims Removed

In addition to the [Chocolatey shortcut shims removed in version 1.0.0](https://github.com/chocolatey/choco/issues/2468), we have removed the [remaining ones](https://github.com/chocolatey/choco/issues/2641).

> The shims removed (inlcuding the earlier ones) and their equivalent full commands are:

Reviewer 2: ```suggestion
The shims removed (including the earlier ones) and their equivalent full commands are:

```

* `cuninst`: use `choco uninstall`.
* `cup`: `choco upgrade`.
* `cver`: `choco version`.

If you want to add these shortcuts back, you can can create a PowerShell function or batch file to provide the same functionality.

## Modified API Key Commands

To make working with Chocolatey CLI consistent, we have changed the `choco apikey` command to add the following subcommands:

* `add`: add an API key.
* `remove`: remove an API key.
* `list`: list the API keys.

As the `remove` subcommand has been added, we have removed the `--remove` options.

## What To Do If Uprading Fails

If you follow the steps below and run into issues that you are unable to resolve, support is available:

* **Licensed Customers**. Please run `choco support` to find the support options available to you.
* **Free and Open-Source (FOSS) Users**. Community support is available on [Discord](https://ch0.co/community).

## Upgrading Chocolatey Products

> :choco-warning: **WARNING**
>
> You **must** upgrade Chocolatey CLI to 1.4.0 before you start your upgrade to the Chocolatey CLI 2.0.0 or Chocolatey Licensed Extension 6.0.0 versions. Failure to do this may result in Chocolatey products being non-functional.

> The Chocolatey products are inter-dependent on each other and all of the products must be upgraded. You cannot upgrade one product and leave the others at older vesions.

Reviewer 2: ```suggestion
The Chocolatey products are inter-dependent on each other and all of the products must be upgraded. You cannot upgrade one product and leave the others at older versions.

```

### Latest 1.x and 5.x Stable Chocolatey Products

To upgrade, you need to ensure you are at the latest stable version of the Chocolatey products which are:

* Chocolatey CLI 1.4.0
* Chocolatey Licensed Extension 5.0.3
* Chocolatey Agent 1.1.2
* Chocolatey GUI 1.1.3
* Chocolatey GUI Licensed Extension 1.0.3

There are two steps to go through, depending on whether you have any of the Chocolatey 2.0.0 or 6.0.0 pre-release products installed.

> <video class="img-fluid" width="320" height="240" poster="/assets/images/chocolatey-upgrade-to-2-banner.png" autoplay muted loop>

Reviewer 2: ```suggestion
<video class="img-fluid" width="800" height="594" poster="/assets/images/chocolatey-upgrade-to-2-banner.png" autoplay muted loop>

```

Here we are adjusting the width and height of that to the video that you took. Doing this makes it much larger and able to read. It was showing very small before. Looks good with these changes!
